### ISSUE TITLE
Cannot silence SC2154 for an associative array index, only for array 

### ISSUE THREAD

**Issue by Contik** [koalaman/shellcheck#1166]
#### For bugs
- Rule Id SC2154
- My shellcheck version 0.4.7 (commit c36f6d89baae1d4e76312ff76f580a8709801b44)
- [X] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

The following snippet works as expected. The script exits if `"${array[zero]}"` is either null or unset. ShellCheck, however, seems to ignore this syntax and keeps reporting [SC2154](https://github.com/koalaman/shellcheck/wiki/SC2154).
```sh
#!/bin/bash
echo "${array[zero]:?'Index is undefined!'}"
```

#### Here's what shellcheck currently says:

```sh
Line 2:
echo "${array[zero]:?'Index is undefined!'}"
      ^-- SC2154: array is referenced but not assigned.
```

I can do as ShellCheck asks and that disables ShellCheck's warning - at the cost of breaking the intended functionality. Everything after `:?` is treated as the error message which is this case will read, `array: Array is undefined![zero]`.
```sh
#!/bin/bash
echo "${array:?'Array is undefined!'[zero]}"
```

#### Here's what I wanted or expected to see:
It'd be nice if ShellCheck allowed silencing this warning for an array index instead of just for an array as a  whole. My current workaround is to use something like `echo "${array:?'Array is undefined!'}"` to silence the warning _*for the array*_, then use `echo "${array[zero]:?'Index is undefined!'}"` where I want it. The second one will still correctly trigger an `exit` if the index is undefined or null, but its syntax is not currently recognized by ShellCheck.

**Comment by matthewpersico** [comment]
Under what conditions is array set? In another script? Exported?

**Comment by Contik** [comment]
In this case it's created in a function `functionName`. Its last command is `declare -p array` which I capture on the calling side via `eval "$(functionName)"`. Thus `"${array[@]}"` and its indices are typically available in my script but not declared there.

I can use "${array:?'Array is undefined!'}" to `exit` when the array is undefined or null, and ShellCheck regocnizes that. Something like `"${array[zero]:?'Index is undefined!'}"` would be neat.

**Comment by koalaman** [comment]
Nicely caught, thanks!

ShellCheck tried to see if it was guarded by whether the text after the variable name started with `:?` or `:-`, which obviously failed for arrays. It should be working now.
